jsEngine/settings/Settings.ts

```diff
@@ -19,19 +27,72 @@ export class JsEnginePluginSettingTab extends PluginSettingTab {
 		this.plugin = plugin;
 	}
 
+	private async listJSfilesInDirectory(directory: string): Promise<string[]> {
+		const { adapter } = this.app.vault;
+		return (await adapter.list(directory)).files.filter(file => file.endsWith('.js'));
+	}
+
 	display(): void {
-		this.containerEl.empty();
+		const { containerEl } = this;
+		const { settings } = this.plugin;
+		containerEl.empty();
 
-		if (!this.plugin.settings) {
+		if (!settings) {
 			return;
 		}
 
 		// this.containerEl.createEl('p', { text: 'Currently Empty, but there will be stuff here later.' });
 
-		new Setting(this.containerEl).setName('Startup scripts').addButton(button => {
+		new Setting(containerEl).setName('Startup scripts').addButton(button => {
 			button.setButtonText('Manage').onClick(() => {
 				new StartupScriptsModal(this.plugin).open();
 			});
 		});
+
+		new Setting(containerEl)
+			.setName('JS snippets (loaded on startup)')
+			.setHeading()
+			.addExtraButton(el => {
+				el.setTooltip('Reload snippets')
+					.setIcon('refresh-cw')
+					.onClick(() => this.display());
+			})
+			.addExtraButton(el => {
+				el.setTooltip('Open snippets folder')
+					.setIcon('folder-open')
+					.onClick(() => this.app.openWithDefaultApp(settings.startupScriptsDirectory));
+			});
+
+		new Setting(containerEl)
+			.setName('JS Snippets Folder')
+			.setDesc('The folder to search for JavaScript files to load')
+			.addText(el => {
+				el.setPlaceholder('Folder')
+					.setValue(settings.startupScriptsDirectory ?? '')
+					.onChange(async (value: string) => {
+						settings.startupScriptsDirectory = value;
+						await this.plugin.saveSettings();
+					});
+			});
+
+		void this.listJSfilesInDirectory(settings.startupScriptsDirectory).then(fileList =>
+			fileList.forEach(file => {
```

Comment:
this feels like there must be a trick to shorten this and only have one level of nesting, instead of pushing the array into a callback that _only_  run `.foreach()` on it with another callback.
 Any tips would be apprecheated.

Reply:
You could refactor this into its own async function. Then you can await the `listJSfilesInDirectory` in there and please use normal loops `for (const x of y)` instead of `.forEach(...)`.